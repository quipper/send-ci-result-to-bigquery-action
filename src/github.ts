export const parseGitHubContext = (contextJson: string): GitHubContext => {
  const context = JSON.parse(contextJson) as unknown
  if (!isGitHubContext(context)) {
    throw new Error(`invalid context JSON: ${contextJson}`)
  }
  return context
}

type GitHubContext = {
  token: string
  run_id: string
}

const isGitHubContext = (x: unknown): x is GitHubContext =>
  typeof x === 'object' && x != null && ('token' in x) && typeof x['token'] === 'string'
