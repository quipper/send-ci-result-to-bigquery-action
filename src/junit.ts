import { XMLParser } from 'fast-xml-parser'
import assert from 'assert'

const parser = new XMLParser({
  removeNSPrefix: true,
  ignoreAttributes: false,
  parseTagValue: true,
  parseAttributeValue: true,
})

export const parseXML = (s: string): TestResult => {
  const xml: unknown = parser.parse(s)
  assertTestResult(xml)
  return xml
}

export type TestResult = {
  testsuite: {
    testcase: {
      '@_classname': string
      '@_name': string
      '@_file': string
      '@_time': number
      failure?: {
        '#text': string
      }
    }[]
  }
}

function assertTestResult(x: unknown): asserts x is TestResult {
  assert(typeof x === 'object')
  assert(x != null)

  assert('testsuite' in x)
  assert(typeof x.testsuite === 'object')
  assert(x.testsuite != null)

  assert('testcase' in x.testsuite)
  assert(Array.isArray(x.testsuite.testcase))
  for (const testcase of x.testsuite.testcase) {
    assertTestcase(testcase)
  }
}

function assertTestcase(x: unknown): asserts x is TestResult['testsuite']['testcase'] {
  assert(typeof x === 'object')
  assert(x != null)
  assert('@_classname' in x)
  assert(typeof x['@_classname'] === 'string')
  assert('@_name' in x)
  assert('@_file' in x)
  assert('@_time' in x)

  if ('failure' in x) {
    assert(typeof x.failure === 'object')
    assert(x.failure != null)
    assert('#text' in x.failure)
    assert(typeof x.failure['#text'] === 'string')
  }
}
