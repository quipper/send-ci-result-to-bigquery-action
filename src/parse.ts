/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { XMLParser } from 'fast-xml-parser'
import assert from 'assert'

const parser = new XMLParser({
  removeNSPrefix: true,
  ignoreAttributes: false,
  parseTagValue: true,
  parseAttributeValue: true,
})

export const parseJUnitXML = (s: string): JUnitXML => {
  const xml = parser.parse(s)
  assertJUnitXML(xml)
  return xml
}

type JUnitXML = {
  testsuite: {
    testcase: {
      '@_classname': string
      '@_name': string
      '@_file': string
      '@_time': number
      failure?: {
        '#text': string
      }
    }[]
  }
}

function assertJUnitXML(x: unknown): asserts x is JUnitXML {
  assert(typeof x === 'object')
  assert(x != null)

  assert('testsuite' in x)
  assert(typeof x.testsuite === 'object')
  assert(x.testsuite != null)

  assert('testcase' in x.testsuite)
  assert(Array.isArray(x.testsuite.testcase))
  for (const testcase of x.testsuite.testcase) {
    assert(typeof testcase === 'object')
    assert(testcase != null)
    assert('@_classname' in testcase)
    assert('@_name' in testcase)
    assert('@_file' in testcase)
    assert('@_time' in testcase)

    if ('failure' in testcase) {
      assert(typeof testcase.failure === 'object')
      assert(testcase.failure != null)
      assert('#text' in testcase.failure)
      assert(typeof testcase.failure['#text'] === 'string')
    }
  }
}
