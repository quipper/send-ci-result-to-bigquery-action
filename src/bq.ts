import * as bigquery from '@google-cloud/bigquery'
import { TestResult, flattenTestCases } from './junit'

type CIResultContext = {
  timestamp: Date
  github_run_id: string
  github_matrix_context_json: string
}

export const parseTestResult = (xml: TestResult, context: CIResultContext) =>
  flattenTestCases(xml).map<CIResultRow>((testcase) => ({
    name: testcase['@_name'],
    classname: testcase['@_classname'],
    file: testcase['@_file'] ?? '',
    time: testcase['@_time'],
    failed: testcase.failure !== undefined,
    failure_message: testcase.failure?.['#text'],
    timestamp: context.timestamp,
    github_run_id: context.github_run_id,
    github_matrix_context_json: context.github_matrix_context_json,
  }))

export type CIResultRow = CIResultContext & {
  name: string
  classname: string
  file: string
  time: number
  failed: boolean
  failure_message: string | undefined
}

export const findOrCreateCiResultTable = async (dataset: bigquery.Dataset, tableId: string) =>
  await findOrCreateTable(dataset, tableId, {
    schema: [
      {
        name: 'timestamp',
        type: 'timestamp',
        mode: 'REQUIRED',
      },
      {
        name: 'name',
        type: 'string',
        mode: 'REQUIRED',
      },
      {
        name: 'classname',
        type: 'string',
        mode: 'REQUIRED',
      },
      {
        name: 'file',
        type: 'string',
        mode: 'REQUIRED',
      },
      {
        name: 'time',
        type: 'numeric',
        mode: 'REQUIRED',
      },
      {
        name: 'failed',
        type: 'boolean',
        mode: 'REQUIRED',
      },
      {
        name: 'failure_message',
        type: 'string',
        mode: 'NULLABLE',
      },
      {
        name: 'github_run_id',
        type: 'string',
        mode: 'REQUIRED',
      },
      {
        name: 'github_matrix_context_json',
        type: 'string',
        mode: 'REQUIRED',
      },
    ],
    timePartitioning: { type: 'DAY', field: 'timestamp' },
  })

export type CIContextRow = {
  timestamp: Date
  failed: boolean
  github_context_json: string
  github_matrix_context_json: string
}

export const findOrCreateCiContextTable = async (dataset: bigquery.Dataset, tableId: string) =>
  await findOrCreateTable(dataset, tableId, {
    schema: [
      {
        name: 'timestamp',
        type: 'timestamp',
        mode: 'REQUIRED',
      },
      {
        name: 'failed',
        type: 'boolean',
        mode: 'REQUIRED',
      },
      {
        name: 'github_context_json',
        type: 'string',
        mode: 'REQUIRED',
      },
      {
        name: 'github_matrix_context_json',
        type: 'string',
        mode: 'REQUIRED',
      },
    ],
    timePartitioning: { type: 'DAY', field: 'timestamp' },
  })

export const insertRowsParallel = async <T>(table: bigquery.Table, rows: T[], batchSize: number) => {
  const promises = []
  for (let head = 0; head < rows.length; head += batchSize) {
    const chunk = rows.slice(head, head + batchSize)
    promises.push(insertRows(table, chunk))
  }
  return await Promise.all(promises)
}

export const insertRows = async <T>(table: bigquery.Table, rows: T[]) => table.insert(rows)

const findOrCreateTable = async (dataset: bigquery.Dataset, tableId: string, options: bigquery.TableMetadata) => {
  const table = dataset.table(tableId)
  const [exists] = await table.exists()
  if (!exists) {
    await dataset.createTable(tableId, options)
  }
  return table
}
