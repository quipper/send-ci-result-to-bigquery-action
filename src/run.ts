import { BigQuery } from '@google-cloud/bigquery'
import { parseGitHubContext } from './github.js'
import * as core from '@actions/core'
import * as bq from './bq.js'
import * as fs from 'fs/promises'
import * as glob from '@actions/glob'
import * as junit from './junit.js'

const BIGQUERY_INSERT_BATCH_SIZE = 100

type Inputs = {
  testResultXMLGlob: string
  bigqueryDatasetName: string
  bigqueryCIResultTableName: string
  bigqueryCIContextTableName: string
  githubContextJSON: string
  githubMatrixContextJSON: string
  googleApplicationCredentialsJSON: string
}

export const run = async (inputs: Inputs): Promise<void> => {
  const timestamp = new Date()
  const githubContext = parseGitHubContext(inputs.githubContextJSON)
  githubContext.token = '***'
  const githubMatrixContext = inputs.githubMatrixContextJSON

  const client = new BigQuery()
  if (inputs.googleApplicationCredentialsJSON) {
    // TODO: should use https://github.com/google-github-actions/auth
    const GOOGLE_APPLICATION_CREDENTIALS_PATH = '/tmp/google_application_credentials.json'
    await fs.writeFile(GOOGLE_APPLICATION_CREDENTIALS_PATH, inputs.googleApplicationCredentialsJSON)
    process.env.GOOGLE_APPLICATION_CREDENTIALS = GOOGLE_APPLICATION_CREDENTIALS_PATH
  }

  core.info(`Creating tables if not exist into dataset ${inputs.bigqueryDatasetName}`)
  const dataset = client.dataset(inputs.bigqueryDatasetName)
  const ciContextTable = await bq.findOrCreateCiContextTable(dataset, inputs.bigqueryCIContextTableName)
  const ciResultTable = await bq.findOrCreateCiResultTable(dataset, inputs.bigqueryCIResultTableName)

  let anyFailed = false
  const xmlGlob = await glob.create(inputs.testResultXMLGlob)
  for await (const xmlPath of xmlGlob.globGenerator()) {
    core.info(`Parsing test result of ${xmlPath}`)
    const xmlContent = await fs.readFile(xmlPath, 'utf-8')
    const testResult = junit.parseXML(xmlContent)
    const testCases = junit.flattenTestCases(testResult)
    const ciResultRows = bq.parseTestResult(testCases, {
      timestamp,
      github_matrix_context_json: githubMatrixContext,
      github_run_id: githubContext.run_id,
    })

    core.info(`Inserting test result of ${xmlPath}`)
    await bq.insertRowsParallel(ciResultTable, ciResultRows, BIGQUERY_INSERT_BATCH_SIZE)
    core.info(`Inserted ${ciResultRows.length} rows`)

    anyFailed = anyFailed || ciResultRows.some((row) => row.failed)
  }

  core.info(`Inserting test context`)
  await bq.insertRows<bq.CIContextRow>(ciContextTable, [
    {
      timestamp,
      github_context_json: JSON.stringify(githubContext),
      github_matrix_context_json: githubMatrixContext,
      failed: anyFailed,
    },
  ])
}
