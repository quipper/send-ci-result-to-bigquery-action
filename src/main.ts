import * as core from '@actions/core'
import { run } from './run'

const main = async (): Promise<void> => {
  await run({
    testResultXMLGlob: core.getInput('test-result-xml-glob', { required: true }),
    bigqueryDatasetName: core.getInput('bigquery-dataset-name', { required: true }),
    bigqueryCIResultTableName: core.getInput('bigquery-ci-result-table-name', { required: true }),
    bigqueryCIContextTableName: core.getInput('bigquery-ci-context-table-name', { required: true }),
    githubContextJSON: core.getInput('github-context-json', { required: true }),
    githubMatrixContextJSON: core.getInput('github-matrix-context-json'),
    googleApplicationCredentialsJSON: core.getInput('google-application-credentials-json'),
  })
}

main().catch((e: Error) => {
  core.setFailed(e)
  console.error(e)
})
