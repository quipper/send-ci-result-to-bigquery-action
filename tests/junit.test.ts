import * as fs from 'fs/promises'
import { assertTestCase, flattenTestCases, parseXML } from '../src/junit'
import { join } from 'path'

describe('rspec.xml', () => {
  const fixturePath = join(__dirname, 'rspec.xml')

  test('parseXML', async () => {
    const xml = parseXML(await fs.readFile(fixturePath, 'utf-8'))
    expect(xml).toBeTruthy()
    expect(xml).toMatchSnapshot()
  })

  test('flattenTestCases', async () => {
    const xml = parseXML(await fs.readFile(fixturePath, 'utf-8'))
    const testCases = flattenTestCases(xml)
    for (const testCase of testCases) {
      assertTestCase(testCase)
    }
    expect(testCases).toMatchSnapshot()
  })
})

describe('jest.xml', () => {
  const fixturePath = join(__dirname, 'jest.xml')

  test('parseXML', async () => {
    const xml = parseXML(await fs.readFile(fixturePath, 'utf-8'))
    expect(xml).toBeTruthy()
    expect(xml).toMatchSnapshot()
  })

  test('flattenTestCases', async () => {
    const xml = parseXML(await fs.readFile(fixturePath, 'utf-8'))
    const testCases = flattenTestCases(xml)
    for (const testCase of testCases) {
      assertTestCase(testCase)
    }
    expect(testCases).toMatchSnapshot()
  })
})
