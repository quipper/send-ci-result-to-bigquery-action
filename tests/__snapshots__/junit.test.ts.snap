// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest.xml flattenTestCases 1`] = `
[
  {
    "@_classname": "parse junit XML",
    "@_name": "parse junit XML",
    "@_time": 0.005,
  },
]
`;

exports[`jest.xml parseXML 1`] = `
{
  "?xml": {
    "@_encoding": "UTF-8",
    "@_version": 1,
  },
  "testsuites": {
    "@_errors": 0,
    "@_failures": 0,
    "@_name": "jest tests",
    "@_tests": 1,
    "@_time": 0.865,
    "testsuite": [
      {
        "@_errors": 0,
        "@_failures": 0,
        "@_name": "undefined",
        "@_skipped": 0,
        "@_tests": 1,
        "@_time": 0.807,
        "@_timestamp": "2023-08-04T12:22:11",
        "testcase": [
          {
            "@_classname": "parse junit XML",
            "@_name": "parse junit XML",
            "@_time": 0.005,
          },
        ],
      },
    ],
  },
}
`;

exports[`rspec.xml flattenTestCases 1`] = `
[
  {
    "@_classname": "spec.features.foo_spec",
    "@_file": "./spec/features/foo_spec.rb",
    "@_name": "foo",
    "@_time": 37.832874,
  },
  {
    "@_classname": "spec.features.bar_spec",
    "@_file": "./spec/features/bar_spec.rb",
    "@_name": "bar",
    "@_time": 32.502537,
  },
  {
    "@_classname": "spec.features.baz_spec",
    "@_file": "./spec/features/baz_spec.rb",
    "@_name": "baz",
    "@_time": 32.297284,
    "failure": {
      "#text": "expected baz",
      "@_message": "example",
      "@_type": "spec.features.baz_spec",
    },
  },
]
`;

exports[`rspec.xml parseXML 1`] = `
{
  "?xml": {
    "@_encoding": "UTF-8",
    "@_version": 1,
  },
  "testsuite": {
    "@_errors": 0,
    "@_failures": 0,
    "@_hostname": "runner",
    "@_name": "rspec",
    "@_skipped": 0,
    "@_tests": 3,
    "@_time": 494.14825,
    "@_timestamp": "2023-08-02T05:15:08+00:00",
    "properties": {
      "property": {
        "@_name": "seed",
        "@_value": 38266,
      },
    },
    "testcase": [
      {
        "@_classname": "spec.features.foo_spec",
        "@_file": "./spec/features/foo_spec.rb",
        "@_name": "foo",
        "@_time": 37.832874,
      },
      {
        "@_classname": "spec.features.bar_spec",
        "@_file": "./spec/features/bar_spec.rb",
        "@_name": "bar",
        "@_time": 32.502537,
      },
      {
        "@_classname": "spec.features.baz_spec",
        "@_file": "./spec/features/baz_spec.rb",
        "@_name": "baz",
        "@_time": 32.297284,
        "failure": {
          "#text": "expected baz",
          "@_message": "example",
          "@_type": "spec.features.baz_spec",
        },
      },
    ],
  },
}
`;
